src/main/java/org/kitteh/irc/client/library/implementation/STSHandler.java

```diff
@@ -0,0 +1,163 @@
+    private final InternalClient client;
+    private final boolean isSecure;
+
+    /**
+     * Creates the event handler for STS.
+     *
+     * @param machine The STS FSM.
+     */
+    public STSHandler(STSMachine machine, InternalClient client) {
+        this.machine = machine;
+        this.client = client;
+        this.isSecure = client.getConfig().getNotNull(Config.SSL);
+    }
+
+    /**
+     * Called when the server responds with its supported capabilities.
+     * @param event The event instance.
+     */
+    @Handler
+    public void onCapLs(CapabilitiesSupportedListEvent event) {
+        // stability not a concern, only one or zero result(s)
+        final Optional<CapabilityState> potentialStsCapability = event.getSupportedCapabilities().stream()
+            .filter(c -> c.getName().equals("sts")).findAny();
+
+        if (!potentialStsCapability.isPresent()) {
+            // get out if we can't do anything useful here
+            return;
+        }
+
+        // okay, we have an STS capability!
+        final CapabilityState sts = potentialStsCapability.get();
+        if (!sts.getValue().isPresent()) {
+            final String msg = event.getOriginalMessages().stream().map(ServerMessage::getMessage)
+                .reduce((a, b) -> (a+b).replace('\n', ' ')).orElse("Missing!");
+            throw new KittehServerMessageException(msg, "No value provided for sts capability.");
+        }
+
+        final String capabilityValue = sts.getValue().get();
+        final Map<String, Optional<String>> options = StringUtil.parseSeparatedKeyValueString(",", capabilityValue);
+        for (String key : options.keySet()) {
+            // Unknown keys are ignored by the switches below
+            this.machine.setSTSPolicy(options);
+            if (this.isSecure) {
+                this.handleSecureKey(key, options);
+            } else {
+                this.handleInsecureKey(key, options);
+            }
+        }
+
+    }
+
+    private void handleInsecureKey(String key, Map<String, Optional<String>> opts) {
+        final Optional<String> value = opts.get(key);
+
+        switch (key) {
+            case "duration":
+                // Do NOT persist, because this policy could've been inserted by an active MitM
+                break;
+            case "port":
+                // Brilliant, we have a secure port. We'll reconnect with SSL and verify the policy.
+                // Quickly ensure the port is valid (less error checking later)
+                if (!value.isPresent()) {
+                    throw new KittehServerMessageException(key, "STS policy had no value for port.");
+                }
+
+                String port = value.get();
+                try {
+                    Integer.parseInt(port); // can't easily use a short because signed..
+                } catch (NumberFormatException nfe) {
+                    throw new KittehServerMessageException(port, "Specified port could not be parsed.");
```

Comment:
I'm personally a fan of attaching the NFE message to the string, but either way's valid.